Add user comment to task instance and dag run

Jorricks opened the pull request:
Implement the support for notes inside Airflow.

Quick showcase:
http://www.giphy.com/gifs/RU8JL1GbYdSic9WiU2
![Giphy](https://media3.giphy.com/media/eEL5HPZJmkFfk5eHe2/giphy.gif?cid=790b761156692a7b34ef545a6f698e12d08d11cfb87a945a&amp;rid=giphy.gif&amp;ct=g)

Closes: #16790

Steps:
1. [x] Add column to database schema.
2. [x] Add code inside DagRun & TaskInstance.
3. [x] Add code to DagRun and TaskInstance views so you can edit it there.
4. [x] Add code inside API to set Task & DagRun notes.
5. [x] Add react code to set the DagRun notes.
6. [x] Make it look & work somewhat decently in the react front-end.
7. [x] Add react code to set the TaskInstances notes.
8. [x] Add minor overlay over DagRuns to indicate there is a note or not.
9. [x] Add minor overlay over TaskInstance to indicate there is a note or not.
10. [x] Make it work with Mapped Instances as well.
11. [x] Implemented autofocus.
12. [x] Implemented accordion to be initially open when there is a note and closed otherwise.
13. [x] Fix all broken tests
14. [x] Added notes support in TriggerDAGRunOperator
15. [x] Add tests for setting notes on DagRun & TaskInstances.
16. [x] Change UserNotes to notes.
17. [x] Create collapsable user notes.
18. [x] Change TaskInstance API endpoint to Patch.
19. [x] Update names of variables.
20. [x] Change corner color to transparent over black.

mik-laj commented on airflow/api_connexion/openapi/v1.yaml:
```diff
@@ -585,6 +585,41 @@ paths:
         '404':
           $ref: '#/components/responses/NotFound'
 
+  /dags/{dag_id}/setTaskInstanceNote:
```
It should be `PATCH /dags/{dag_id}/`. This will allow several changes to be made with a single HTTP request.

Jorricks replied:
Indeed, should have made this a PATCH operation.
However, this ties in closer with a DAGRun than a DAG in general IMO.
Shall I make it part of the already existing patch operation in `/dags/{dag_id}/dagRuns/{dag_run_id}:`?
https://github.com/apache/airflow/blob/527c90c9804ee66cfa800381cbd67280e5d3a8a9/airflow/api_connexion/openapi/v1.yaml#L787

Jorricks replied:
Or even better probably, make a patch operation on this endpoint?
 `/dags/{dag_id}/dagRuns/{dag_run_id}/taskInstances/{task_id}:`
https://github.com/apache/airflow/blob/527c90c9804ee66cfa800381cbd67280e5d3a8a9/airflow/api_connexion/openapi/v1.yaml#L1193

bbovenzi replied:
We'd want both `/dags/{dag_id}/dagRuns/{dag_run_id}` and `/dags/{dag_id}/dagRuns/{dag_run_id}/taskInstances/{task_id}` right?

I would've suggested extending `/dags/{dag_id}/updateTaskInstancesState` but the url name doesn't really work.

Jorricks replied:
Yes to me that makes sense. I'll change it and see if I can get somewhere with the front-end soon. Thanks for the early feedback

Jorricks replied:
I changed this. Marking this as resolved

bbovenzi commented on airflow/www/static/js/dag/grid/dagRuns/Bar.tsx:
```diff
@@ -110,7 +110,11 @@ const DagRunBar = ({
-            backgroundColor={stateColors[state]}
+            background={
+              run.notes != null && run.notes !== ''
+                ? `linear-gradient(-135deg,#000 8px, ${stateColors[state]} 0);`
```
Could you provide more screenshots of what a dag run / task instance with a note looks like?
I'm not sure if we should have a visual indication on the grid if something has a note.

Jorricks replied:
I think we should have a visual indication (as otherwise in a list of 100 TaskInstances, you'd never find the note).
Anyways, I think we can still tune it a bit. I don't have great designer skills, but thought this looks reasonable and is very distinctive. We can also change the color, figured black would just fit easily on top of all the bright colors we have.
<img width="1338" alt="image" src="https://user-images.githubusercontent.com/20257392/192553634-7ade2212-ae12-4b91-95d6-d761b0f8903c.png">

Jorricks replied:
Good pointer. I thought that was only for non-API endpoints but I see it is now.
Changing this now.

Jorricks replied:
Also added support for this in the Mapped instances. Don't mind the Chrome black inversion :)
<img width="775" alt="image" src="https://user-images.githubusercontent.com/20257392/192575928-d3bec207-cdfc-4e61-83f3-9008141679a9.png">

bbovenzi replied:
You're right, there's probably a reason for a note. Idk about the black gradient, but I'll think more on what else could work.

bbovenzi commented on airflow/www/static/js/dag/details/taskInstance/index.tsx:
```diff
@@ -90,6 +92,8 @@ const TaskInstance = ({
 
   if (!group || !run || !instance) return null;
 
+  const updateTaskInstanceNotesCallable = (newNotes: string) => { instance.notes = newNotes; };
```
We shouldn't mutate instance like this.

Let's keep notes as a state variable with
```const [notes, setNotes] = useState(instance.notes || '');```

And then here we can call `setNotes(newNotes);`

bbovenzi replied:
oh wait, you do that in `SetDagTaskNotes.tsx`. What is this doing?

Jorricks replied:
Yes so the idea in both cases is that we update the data we fetched from the API.
The reason for that is that when we switch DagRuns/TaskInstance views and go back, that it will still show you the most up to date task note. I figured we could either reload the entire dataset, or we could simply update the data we fetched, so I went for the second option.
How would you approach this issue?

bbovenzi replied:
Ahh yes, we should add an `onSuccess` callback to the set notes API functions. In that function we get the grid data and find the instance in question and update its note. Unfortunately, that is quite a roundabout way to do it, but it ensures that the data is properly updated.

I can implement that part if you want.

Jorricks replied:
I very much like that idea. Much less hacky.
 Let me give that a try :)

bbovenzi replied:
You can look at the [docs](https://react-query-v3.tanstack.com/guides/optimistic-updates) for optimistic updates for an example of how to get and set query data.

Jorricks replied:
I updated it. Took me ages to get it right.
Let me know what you think.
IMO it's much much better then what I originally had, so thanks a lot for raising this :)

bbovenzi commented on airflow/www/static/js/api/useSetDagRunNotes.ts:
```diff
@@ -0,0 +1,40 @@
+  newNotesValue: string,
+) {
+  const errorToast = useErrorToast();
+  const setDagRunNotes = (`/api/v1/dags/${dagId}/dagRuns/_DAG_RUN_ID_/setNote`)
```
We can't hardcode this value. Take a look at the other API files for an example of how we do it. You'll need to make a change to `grid.html` too.

Jorricks replied:
Added now. Give the pre-commit a couple minutes to run though xD

bbovenzi commented on airflow/www/static/js/api/useSetTaskInstanceNotes.ts:
```diff
@@ -0,0 +1,40 @@
+  newNotesValue: string,
+) {
+  const errorToast = useErrorToast();
+  const setTaskInstanceNotes = `/api/v1/dags/${dagId}/dagRuns/${runId}/taskInstances/${taskId}/setNote`;
```
Let's also use meta tags for this url too.

Jorricks replied:
Added now. Give the pre-commit a couple minutes to run though xD

bbovenzi commented on airflow/www/static/js/dag/details/SetDagTaskNotes.jsx:
```diff
@@ -0,0 +1,120 @@
+    <Accordion defaultIndex={[0]} allowMultiple>
+      <AccordionItem style={{ border: 0 }}>
+        <AccordionButton style={{ padding: 0, paddingBottom: '10px', fontSize: 'inherit' }}>
+          <Box flex="1" textAlign="left">
```
Btw, if you want to simplify this. Chakra has a pre-existing component we can leverage: https://chakra-ui.com/docs/components/editable

Jorricks replied:
I checked this page out some more but I really like the automatic extending text area.
It's a much smoother experience IMO then setting 3 lines and forcing someone to stick to the 3 lines.
So I'd prefer to keep it as it is.

bbovenzi replied:
Cool. That's fine. I just wanted to make sure you knew there was something available.

pierrejeambrun replied:
I am not a big fan of adding a new dependency just for the auto resize feature of the textarea. Chakra also provides its own textarea [component](https://chakra-ui.com/docs/components/textarea). it is possible to add the autoresize behavior:
https://github.com/chakra-ui/chakra-ui/issues/670#issuecomment-669916624

Is it possible to convert this file to typescript ? (Or is there a specific reason for the `.jsx`?)

Jorricks replied:
I don't mind changing it to the Chakra component with the autoresize. Just note that that autoresize behaviour is also based on the same library.

Regarding `.jsx`, I am completely new to the `node.js` ecosystem. I created a `.jsx` because the most similar item also was `.jsx`. Should I try to make it a `.tsx`?

Jorricks replied:
Moved it onto a `.tsx`. This forced me to restructure some parts and highlighted some typing improvements. Thanks for the suggestion @pierrejeambrun! Let me know if I can improve anything else :)

pierrejeambrun replied:
Seems good 👍 thanks.

I’ll do a full review later today :)

bbovenzi commented on airflow/www/static/js/api/useSetDagRunNotes.ts:
```diff
@@ -0,0 +1,40 @@
+import { useMutation } from 'react-query';
+import useErrorToast from '../utils/useErrorToast';
+
+export default function UseSetDagRunNotes(
```
```suggestion
export default function useSetDagRunNotes(
```

Jorricks replied:
Woops. Changed :)

bbovenzi commented on airflow/www/static/js/dag/details/taskInstance/Details.tsx:
```diff
@@ -24,7 +24,8 @@ import {
   Tr,
-  Td, Heading, Thead,
+  Td,
+  Thead,
```
I think this file has been updated. You may want to rebase.

Jorricks replied:
Yes there is a big rebase coming in a bit I think xD

bbovenzi commented on airflow/www/static/js/dag/details/taskInstance/MappedInstances.tsx:
```diff
@@ -63,7 +63,11 @@ const MappedInstances = ({
+        <StatusWithNotes
+          state={mi.state === undefined || mi.state === 'none' ? null : mi.state}
+          mx={2}
+          containsNotes={mi.notes != null && mi.notes !== ''}
```
```suggestion
          containsNotes={!!mi.notes}
```

bbovenzi commented on airflow/www/static/js/dag/grid/renderTaskRows.tsx:
```diff
@@ -95,6 +95,7 @@ const TaskInstances = ({
                 group={task}
                 onSelect={onSelect}
                 isActive={hoveredTaskState === undefined || hoveredTaskState === instance.state}
+                containsNotes={instance.notes != null && instance.notes !== ''}
```
```suggestion
                containsNotes={!!instance.notes}
```

Jorricks committed:
Implement Notes for TaskInstances and DagRuns

pierrejeambrun commented on airflow/api_connexion/endpoints/task_instance_endpoint.py:
```diff
@@ -545,3 +546,54 @@ def post_set_task_instances_state(*, dag_id: str, session: Session = NEW_SESSION
+    ],
+)
+@provide_session
+def set_task_instance_note(
```
```suggestion
def set_task_instance_notes(
```

I think the name should match the field we set.

Jorricks replied:
Changed :)

pierrejeambrun commented on airflow/api_connexion/endpoints/dag_run_endpoint.py:
```diff
@@ -413,3 +414,28 @@ def clear_dag_run(*, dag_id: str, dag_run_id: str, session: Session = NEW_SESSIO
+    ],
+)
+@provide_session
+def set_dag_run_note(*, dag_id: str, dag_run_id: str, session: Session = NEW_SESSION) -> APIResponse:
```
```suggestion
def set_dag_run_notes(*, dag_id: str, dag_run_id: str, session: Session = NEW_SESSION) -> APIResponse:
```
Same

Jorricks replied:
Changed :)

pierrejeambrun commented on airflow/api_connexion/endpoints/task_instance_endpoint.py:
```diff
@@ -545,3 +546,54 @@ def post_set_task_instances_state(*, dag_id: str, session: Session = NEW_SESSION
+@security.requires_access(
+    [
+        (permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG),
+        (permissions.ACTION_CAN_EDIT, permissions.RESOURCE_DAG_RUN),
```
Here we are updating a TaskInstance, should we also require the permission for that  ?
```python
(permissions.ACTION_CAN_EDIT, permissions.RESOURCE_TASK_INSTANCE)
```

Jorricks replied:
Good one! Completely missed that. Added 👍

pierrejeambrun commented on airflow/api_connexion/endpoints/task_instance_endpoint.py:
```diff
@@ -545,3 +546,54 @@ def post_set_task_instances_state(*, dag_id: str, session: Session = NEW_SESSION
+    session.commit()
+
+    if map_index == -1:
+        return get_task_instance(dag_id=dag_id, dag_run_id=dag_run_id, task_id=task_id)
```
We already have the TaskInstance available in the scope,maybe directly serializing this instance and return it would simplify and avoid the extra db query?
```python
    return task_instance_schema.dump((ti, None)) # The task_instance_schema is a bit special :)
```

If we need the `SlaMiss` and `rendered_task_instance_fields` you can add them to your default query and update the dump call.

Jorricks replied:
Yes I tried this but I couldn't get this to work. The moment I add the joins I get a `Row` object back instead of an ORM `TaskInstance` object. Therefore, I have the feeling they need to be two separate queries?
The error I am seeing is;
```
    ti.notes = new_value_for_notes
    session.commit()
    return task_instance_schema.dump(ti)
AttributeError: 'Row' object has no attribute 'notes'
```
This was the query;
```

    query = (
        session.query(TI)
        .filter(TI.dag_id == dag_id)
        .filter(TI.run_id == dag_run_id)
        .filter(TI.task_id == task_id)
    )
    if map_index == -1:
        query = query.filter(or_(TI.map_index == -1, TI.map_index is None))
    else:
        query = query.filter(TI.map_index == map_index)

    query = (
        query
        .join(TI.dag_run)
        .outerjoin(
            SlaMiss,
            and_(
                SlaMiss.dag_id == TI.dag_id,
                SlaMiss.execution_date == DR.execution_date,
                SlaMiss.task_id == TI.task_id,
            ),
        )
        .add_entity(SlaMiss)
        .options(joinedload(TI.rendered_task_instance_fields))
    )
```
Maybe I am missing something. Could you give it a try perhaps?

pierrejeambrun replied:
I tested the line I a have put above and it worked. The difference is to pass a tuple to the dump call. (The task instance schema is different from what we may expect, in a sense that it does need more than just the TI for serialising).

For your specific error, (the add_entity) change your result to a tuple, so you need to take that into account if you update your query like that, you don’t necessarily need that. For your particular error, you cannot do query.all().notes, because the return of query.all(), is not a TI anymore. (add_entity added the SlaMiss in the result set).


```python
    query = (
        session.query(TI)
        .filter(TI.dag_id == dag_id)
        .filter(TI.run_id == dag_run_id)
        .filter(TI.task_id == task_id)
    )
    if map_index == -1:
        query = query.filter(or_(TI.map_index == -1, TI.map_index is None))
    else:
        query = query.filter(TI.map_index == map_index)

    try:
        ti: TI | None = query.one_or_none()
    except MultipleResultsFound:
        raise NotFound(
            "Task instance not found", detail="Task instance is mapped, add the map_index value to the URL"
        )
    if ti is None:
        error_message = f'Task Instance not found for dag_id={dag_id}, run_id={dag_run_id}, task_id={task_id}'
        raise NotFound(error_message)

    ti.notes = new_value_for_notes
    session.commit()

    return task_instance_schema.dump((ti, None))
```

pierrejeambrun replied:
If you want to use the same query as in `get_task_instance`:
```python
    query = (
        session.query(TI)
        .filter(TI.dag_id == dag_id, TI.run_id == dag_run_id, TI.task_id == task_id)
        .join(TI.dag_run)
        .outerjoin(
            SlaMiss,
            and_(
                SlaMiss.dag_id == TI.dag_id,
                SlaMiss.execution_date == DR.execution_date,
                SlaMiss.task_id == TI.task_id,
            ),
        )
        .add_entity(SlaMiss)
        .options(joinedload(TI.rendered_task_instance_fields))
    )
    if map_index == -1:
        query = query.filter(or_(TI.map_index == -1, TI.map_index is None))
    else:
        query = query.filter(TI.map_index == map_index)

    try:
        ti, sla_miss = query.one_or_none()
    except MultipleResultsFound:
        raise NotFound(
            "Task instance not found", detail="Task instance is mapped, add the map_index value to the URL"
        )
    if ti is None:
        error_message = f'Task Instance not found for dag_id={dag_id}, run_id={dag_run_id}, task_id={task_id}'
        raise NotFound(error_message)

    ti.notes = new_value_for_notes
    session.commit()

    return task_instance_schema.dump((ti, sla_miss))

```

Jorricks replied:
That works perfectly! Thanks. I completely missed the unpacking that happened.

pierrejeambrun commented on airflow/www/static/js/api/useSetDagRunNotes.ts:
```diff
@@ -0,0 +1,61 @@
+
+const setDagRunNotesURI = getMetaValue('set_dag_run_note');
+
+export default function useSetDagRunNotes(
```
Types are auto generated for the API. This way typescript types are always in sync with the api spec. (allows us to write less code and spot bugs much more easily). For instance in this case you can directly use:
```
API.SetDagRunNoteVariables
```
You can take a look at `useTaskInstance` hook to get an example.

Could directly write something like:

![image](https://user-images.githubusercontent.com/14861206/192875363-9d041bdf-fc35-4aa6-805f-95a00fe8c6f3.png)

Jorricks replied:
This worked for SetDagRunNote, but it doesn't work for items that require a `body` because there seems to be a bug that marks these fields as optional even though I put that the `body` is required.

```
      requestBody:
        description: Parameters of set Task Instance note.
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/SetTaskInstanceNote'
```
Leads to
```
    SetTaskInstanceNote: {
      /** @description The map_index of the TaskInstance. */
      map_index?: number;
      /** @description The custom note to set for this Task Instance. */
      notes?: string;
    };
```

pierrejeambrun replied:
By default all object properties are optional, you need to mark them as `required` if needed, for instance here, in the schema, add the 'required' field.
```yaml
    SetTaskInstanceNote:
      type: object
      required:
        - map_index
      properties:
        map_index:
          description: The map_index of the TaskInstance.
          type: integer
        notes:
          description: The custom note to set for this Task Instance.
          type: string
```

will produce:
```ts
    SetTaskInstanceNote: {
      /** @description The map_index of the TaskInstance. */
      map_index: number;
      /** @description The custom note to set for this Task Instance. */
      notes?: string;
    };

```

Jorricks replied:
I wasn't aware of this. Thanks!!

pierrejeambrun commented on airflow/www/static/js/api/useSetDagRunNotes.ts:
```diff
@@ -0,0 +1,61 @@
+    if (oldValue == null) return emptyGridData;
+    const run = oldValue.dagRuns.find((dr) => dr.runId === runId);
+    if (run) {
+      run.notes = newNotesValue;
```
I'm not really sure about mutations like this, as we usually try to avoid it, but I don't see other options for this particular case. (maybe it's fine as is :thinking:. I have no idea to be honest)

Jorricks replied:
This is what @bbovenzi recommended I believe. Not sure if this was exactly what he had in mind as well?

bbovenzi replied:
No, we still shouldn't mutate run like that. I'll add a suggestion

bbovenzi replied:
Fixed!

pierrejeambrun commented on airflow/www/static/js/api/useSetDagRunNotes.ts:
```diff
@@ -0,0 +1,61 @@
+      onSuccess: async () => {
+        // Cancel any outgoing refetches (so they don't overwrite our optimistic update)
+        await queryClient.cancelQueries('gridData');
+        // Optimistically update to the new value
```
Really interesting :+1:

pierrejeambrun commented on airflow/www/static/js/api/useSetTaskInstanceNotes.ts:
```diff
@@ -0,0 +1,107 @@
+) {
+  const queryClient = useQueryClient();
+  const errorToast = useErrorToast();
+  // Note: Werkzeug does not like the META URL with an integer. It can not put _MAP_INDEX_ there
```
cf my comment to maybe try to avoid this, and `useTaskInstance` handle this same scenario. (maybe not in the best way but might be worth to take a look)

**edit:** The comment I am referring to never made it through apparently :).

For the `get_task_instance` and `get_mapped_task_instance`, we have the same problem. What we do is  that we pass the trailing `/map_index` for mapped task, so the map_index is defined and >=0. For unmapped task we just request without the trailing map index, such as:
- /dags/{dag_id}/dagRuns/{dag_run_id}/taskInstances/{task_id} -> for unmapped task
- /dags/{dag_id}/dagRuns/{dag_run_id}/taskInstances/{task_id}/{map_index} -> for mapped tasks

This way we never pass -1 in the path. This is for a GET method where passing it in the body isn't really an option (even if not prohibited by the spec :joy:), we have more freedom here. But maybe this forces us to have 2 different entries in the openapi spec... :thinking:

Jorricks replied:
Yes I had some doubts here as well.. 
I felt it's not a good approach to now start splitting up each TaskInstance related endpoint into two. That would lead to code duplication in many places, which would be a shame.
I felt this was the best way of handling it.

pierrejeambrun replied:
Just an idea but maybe those 2 entries in the spec can map to the same `x-openapi-router-controller` to not have code duplication.

Jorricks replied:
Ohhh nice idea!!! I will give that a try :)!

Jorricks replied:
Managed to make it work somewhat. Had to still introduce a different function as it wouldn't be able to show the different URLs otherwise.

pierrejeambrun replied:
Nice, let’s see what others think, maybe your original idea of passing it in the body is more suitable in this case :) (having to handle 2 meta tags in the front might be not ideal. I didn’t anticipate that)

bbovenzi replied:
Hmm. Yeah, if integers don't work in the URL then I guess putting it in the body makes sense.

Jorricks replied:
I think either is fine but this is more in line with the getTaskInstance information endpoints. Is it okay to keep is as it is (considering this doesn't need to be set in stone of course :))?

pierrejeambrun commented on airflow/www/static/js/api/useSetTaskInstanceNotes.ts:
```diff
@@ -0,0 +1,107 @@
+  runId: string,
+  taskId: string,
+  mapIndex: number,
+  newNotesValue: string,
```
Same here you should be able to directly use the API generated type. `SetTaskInstanceNoteVariables`

Jorricks replied:
See message above.

pierrejeambrun reviewed (commented):
Very cool feature, I tried that locally and this is very exciting !

I just left a few comments, hope this helps.

pierrejeambrun commented on airflow/www/static/js/dag/details/SetDagTaskNotes.tsx:
```diff
@@ -0,0 +1,134 @@
+}: Props) => {
+  const [notes, setNotes] = useState(initialValue == null ? '' : initialValue);
+  const [noteBeforeEdit, setNoteBeforeEdit] = useState('');
+  const [editMode, changeEditMode] = useState(false);
```
nit: just by convention
```suggestion
  const [editMode, setEditMode] = useState(false);
```

Jorricks replied:
Changed :)

pierrejeambrun commented on airflow/www/static/js/dag/details/SetDagTaskNotes.tsx:
```diff
@@ -0,0 +1,134 @@
+  };
+
+  return (
+    <Accordion defaultIndex={[0]} allowMultiple>
```
```suggestion
    <Accordion allowToggle>
```
We only have 1 AccordionItem, I don't think the `allowMultiple` is necessary in that case. [allow multiple ](https://chakra-ui.com/docs/components/accordion#expand-multiple-items-at-once).

`allowToggle` might be what you are looking for ?

Jorricks replied:
You are right :)! Should be better now

pierrejeambrun commented on airflow/www/static/js/dag/details/SetDagTaskNotes.tsx:
```diff
@@ -0,0 +1,134 @@
+              </div>
+              <div style={{ marginTop: '10px' }}>
+                <Button type="submit" isLoading={dagRunIsLoading || tiIsLoading}>
+                  Update user notes
```
Looks like almost all buttons in UI are in Title Case -> `Update User Notes`. Maybe we should update these new buttons to harmonize with the rest of the UI ?

Jorricks replied:
Yes indeed. Changed

Jorricks committed:
FIX Add simple autoFocus

Jorricks committed:
FIX Open/close accordion accordingly.

Jorricks committed:
Update according to comments

Jorricks committed:
FIX Tune UI according to comments

Jorricks committed:
FIX Add separate endpoint for mapped tasks

Jorricks committed:
FIX Add property as required

Jorricks committed:
FIX Set notes back to None when empty string

Jorricks committed:
FIX All broken tests

Jorricks committed:
FIX Flaky test

Jorricks committed:
FIX Add notes as option to TriggerDagRunOperator

bbovenzi commented on airflow/www/static/js/api/useSetDagRunNotes.ts:
```diff
@@ -0,0 +1,61 @@
+
+  const updateGridData = (oldValue: GridData | undefined) => {
+    if (oldValue == null) return emptyGridData;
+    const run = oldValue.dagRuns.find((dr) => dr.runId === runId);
```
```suggestion
    return oldValue.dagRuns.map((dr) => {
       if (dr.runId === runId) {
         return {
           ...dr,
           notes: newNotesValues,
         }
       }
       return dr;
    });
```

bbovenzi replied:
This will update the cache without mutating the existing data.

bbovenzi replied:
I added a commit to fix this.

bbovenzi commented on airflow/www/static/js/api/useSetDagRunNotes.ts:
```diff
@@ -0,0 +1,61 @@
+  const errorToast = useErrorToast();
+  const setDagRunNotes = setDagRunNotesURI.replace('_DAG_RUN_ID_', runId);
+
+  const updateGridData = (oldValue: GridData | undefined) => {
```
```suggestion
  const updateGridData = (oldGridData: GridData | undefined) => {
```
Let's have a more descriptive name?

Jorricks replied:
When I changed the name, I got a syntax error. So it seems, my IDE does not like me changing that.

bbovenzi replied:
I figured it out. This should be fixed now.

bbovenzi commented on airflow/www/static/js/api/useSetTaskInstanceNotes.ts:
```diff
@@ -0,0 +1,107 @@
+    const group = getTask({ taskId, task: oldValue.groups });
+    const instance = group?.instances.find((ti) => ti.runId === runId);
+    if (instance) {
+      instance.notes = newNotesValue;
```
See my suggested comment on useSetDagRunNotes

bbovenzi commented on airflow/www/static/js/api/useSetTaskInstanceNotes.ts:
```diff
@@ -0,0 +1,107 @@
+  const updateMappedInstancesResult = (oldValue: API.TaskInstanceCollection | undefined) => {
+    if (oldValue == null) {
+      return {
+        taskInstances: undefined,
```
Let's default to an empty array.
```suggestion
        taskInstances: [],
```

bbovenzi commented on airflow/www/static/js/api/useTaskInstance.ts:
```diff
@@ -55,7 +55,7 @@ const useTaskInstance = ({
 
   return useQuery(
-    ['taskIntance', dagId, dagRunId, taskId, mapIndex],
+    ['taskInstance', dagId, dagRunId, taskId, mapIndex],
```
Good catch!

bbovenzi commented on airflow/www/static/js/dag/StatusBox.tsx:
```diff
@@ -36,15 +36,32 @@ import InstanceTooltip from './InstanceTooltip';
+  containsNotes?: boolean;
 }
+export const StatusWithNotes = ({ state, containsNotes, ...rest }: StatusWithNotesProps) => {
+  const theColor = state && stateColors[state] ? stateColors[state] : 'white';
```
Can we get a more descriptive name please?

Jorricks replied:
Changed `theColor` to `backgroundColor`

Jorricks committed:
Add tests for API endpoints

Jorricks committed:
FIX Broken test

bbovenzi committed:
fix query cache updates

bbovenzi committed:
add react tests, fix warnings, improve api mutations

Jorricks committed:
Update colors in the TI StatusBox & DagRun Bar

bbovenzi committed:
improve note marker on statuses and note action buttons

bbovenzi committed:
indicate note in run/task tooltip

Jorricks committed:
FIX Yarn incorrectly updated during rebase

bbovenzi committed:
fix typescript errors

bbovenzi committed:
fix migration files

bbovenzi committed:
fix rebase

Jorricks committed:
FIX api_connexion tests

Jorricks committed:
FIX www tests

ephraimbuddy commented on airflow/models/taskinstance.py:
```diff
@@ -452,6 +452,7 @@ class TaskInstance(Base, LoggingMixin):
     queued_by_job_id = Column(Integer)
     pid = Column(Integer)
     executor_config = Column(ExecutorConfigType(pickler=dill))
+    notes = Column(String(1000))
```
```suggestion
    notes = Column(String(1000).with_variant(Text(1000),'mysql'))
```
Without this, ORM created DB will have String for MySQL.

Jorricks replied:
@ephraimbuddy do you have a reference somewhere on why this is important? I am unfamiliar with this problem and would like to get more familiar :).

ephraimbuddy replied:
About with_variant: https://docs.sqlalchemy.org/en/14/core/type_api.html#sqlalchemy.types.TypeEngine.with_variant

Let me explain why we need this change.

In your migration file, you have an `else` statement to use `Text` or `String` depending on whether the DB is MySQL or not. However, the ORM is just `String`. You can find out the column type created in MySQL when the DB is created from ORM by running this command `airflow db init` on a clean DB and compare the column type created when you create the DB from migration file by running `airflow db upgrade -r heads` on a clean DB as well. 

Both ORM column types and migration files should be the same to avoid issues with column change in the future.

The test is passing because we ignored String columns in type compare see: https://github.com/apache/airflow/blob/8f152a102baf6851a719e4ad10fbb15e928fa72b/airflow/utils/db.py#L1796-L1802

Another place we have done a similar thing is https://github.com/apache/airflow/blob/8f152a102baf6851a719e4ad10fbb15e928fa72b/airflow/models/connection.py#L92. I believe there are other places too. 

People creating DB from ORM should have the same column types as those creating from migration file so we don't have issue when changing the type in the future

ephraimbuddy commented on airflow/models/dagrun.py:
```diff
@@ -119,6 +119,7 @@ class DagRun(Base, LoggingMixin):
     # When a scheduler last attempted to schedule TIs for this DagRun
     last_scheduling_decision = Column(UtcDateTime)
     dag_hash = Column(String(32))
+    notes = Column(String(1000))
```
```suggestion
    notes = Column(String(1000).with_variant(Text(1000),'mysql'))
```

ephraimbuddy commented on airflow/www/static/js/api/useDatasetEvents.ts:
```diff
@@ -30,7 +30,7 @@ export default function useDatasetEvents({
     ['datasets-events', datasetId, sourceDagId, sourceRunId, sourceTaskId, sourceMapIndex, limit, offset, orderBy],
     () => {
-      const datasetsUrl = getMetaValue('dataset_events_api') || '/api/v1/datasets/events';
+      const datasetsUrl = getMetaValue('dataset_events_api');
```
Why remove this?

Jorricks replied:
Because we get the urls from the `html` file using the getMetaValue. 

If these are not working, there is no reason for this single endpoint to keep working while all others (that don't have this 'else' in there) fail. Also, now we need to maintain the API route on two locations instead of on one location, so it's neater to remove it IMO.

bbovenzi replied:
Yeah, `getMetaValue()` was updated to handle `undefined` better. This is fine.

ashb replied:
Turns out that the grid.html was missing this value.

And due to how `getMetaValue` currently behaves (it just returns an empty string!) this just showed up as "no dataset events" instead.

Temporary fix https://github.com/apache/airflow/pull/27927 but we should go and make getMetaValue error if the requested tag isn't found (but lets do that _after_ 2.5.)

@Jorricks For future: please try to keep entirely unrelated changes like this to a separate PR (this change has 0 to do with notes.)

Jorricks replied:
Noted. Apologies, next time I'll make sure to move unrelated (thought-to-be) improvements in separate PRs. 
Thanks for getting back to me about this and I hope you didn't spend to much time on this issue.

Jorricks committed:
Apply suggestions from code review

Use with_variant to ensure db init and db upgrade lead to same types.
Thanks @ephraimbuddy

Co-authored-by: Ephraim Anierobi <splendidzigy24@gmail.com>

Jorricks committed:
FIX missing import

bbovenzi reviewed (approved):
UI-wise LGTM!

pierrejeambrun commented on airflow/www/static/js/api/useSetDagRunNotes.ts:
```diff
@@ -0,0 +1,67 @@
+
+  return useMutation(
+    ['setDagRunNotes', dagId, runId],
+    (notes: string | null) => axios.patch(setDagRunNotes, { notes }),
```
```suggestion
    (notes: string | null) => axios.patch<AxiosResponse, API.DagRun>(setDagRunNotes, { notes }),
```

pierrejeambrun commented on airflow/www/static/js/api/useSetDagRunNotes.ts:
```diff
@@ -0,0 +1,67 @@
+    (notes: string | null) => axios.patch(setDagRunNotes, { notes }),
+    {
+      onSuccess: async (data) => {
+        const notes = (data as API.DAGRun).notes ?? null;
```
So you can avoid casting here
```suggestion
        const notes = data.notes ?? null;
```

pierrejeambrun commented on airflow/www/static/js/api/useSetDagRunNotes.ts:
```diff
@@ -0,0 +1,67 @@
+ * under the License.
+ */
+
+import axios from 'axios';
```
```suggestion
import axios, { AxiosResponse } from 'axios';
```

pierrejeambrun commented on airflow/www/static/js/api/useSetDagRunNotes.ts:
```diff
@@ -0,0 +1,67 @@
+interface Props {
+  dagId: string;
+  runId: string;
+}
```
For late: Too bad we cannot use `SetDagRunNotesVariables` to type this. I think we should separate the type generation for `parameters/path & query` and `requetsBody`.

This way we could easily type such mutation, keeping type enforced by the OpenAPI spec. What do you think @bbovenzi

pierrejeambrun replied:
For instance the `notes`  type would raise a type error I believe, it can be `undefined | string` but not `null | string` I believe.

pierrejeambrun commented on airflow/www/static/js/api/useSetTaskInstanceNotes.ts:
```diff
@@ -0,0 +1,114 @@
+    (notes: string | null) => axios.patch(url, { notes }),
+    {
+      onSuccess: async (data) => {
+        const notes = (data as API.TaskInstance).notes ?? null;
```
Same here to avoid the `as`

pierrejeambrun commented on airflow/www/static/js/api/useSetTaskInstanceNotes.ts:
```diff
@@ -0,0 +1,114 @@
+          updateTaskInstanceResult,
+        );
+      },
+      onError: (error: Error) => errorToast({ error }),
```
Shouldn't we rollback the optimistic update in case there is an error ?

bbovenzi replied:
This isn't an optimistic update. It only runs `onSuccess`. If it was `onMutate`, that would be an optimistic update.

pierrejeambrun replied:
You are right thanks

ashb reviewed (commented):
Why did we need to change the "configure-aws-credentials" action submodule in this PR?

pierrejeambrun commented on airflow/www/static/js/utils/colors.json:
```diff
@@ -0,0 +1,842 @@
+[
```
I don't like having to add an 800 json file to do color mapping just for the note linear-gradient. I wish we could do without it. Or implement a solution that leverage ChakraUI themes and color palette. (We could for instance only map default colors to chakra color palette and then play with the numbers to get brighter colors -> red.400, red.200 etc.)

bbovenzi replied:
Status colors are not defined in the UI and are stored as strings "green" instead of rgb or hex. I think it would be best to maintain a single source of truth for state colors. I wanted a way to convert it. But using [an existing library](https://www.npmjs.com/package/color) is probably better.

pierrejeambrun replied:
Very interesting. Maybe for later, I can do that.

pierrejeambrun reviewed (approved):
Overall LGTM (a bit scary indeed :joy:)

I just made a few suggestions but non blocking.

Out of curiosity, I see changes to the submodule `.github/actions/configure-aws-credentials`. Is it normal, where does this come from ?

Jorricks committed:
[PATCH] address PR UI feedback
(by Brent Bovenzi)

Jorricks committed:
FIX Make static checks happy
